dali/operators/crop/slice_attr.h

```diff
@@ -19,105 +19,129 @@
-            // Assumes xywh
-            ProcessArgumentsHelper(
-                data_idx,
-                crop_size.data<float>()[0],
-                crop_size.data<float>()[1],
-                crop_begin.data<float>()[0],
-                crop_begin.data<float>()[1]);
-        }
+  void ProcessArguments(DeviceWorkspace &ws) {
+    DALI_ENFORCE(ws.NumInput() == 3,
+      "Expected 3 inputs. Received: " + std::to_string(ws.NumInput()));
+    const auto &crop_anchor = ws.Input<CPUBackend>(1);
+    const auto &crop_shape = ws.Input<CPUBackend>(2);
+    for (std::size_t data_idx = 0; data_idx < batch_size__; data_idx++) {
+      VerifyArgsShape(crop_anchor.tensor_shape(data_idx), crop_shape.tensor_shape(data_idx));
+      ProcessArgumentsHelper(data_idx, crop_anchor.tensor<float>(data_idx),
+                             crop_shape.tensor<float>(data_idx));
     }
+  }
+
+  void ProcessArguments(const SampleWorkspace &ws) {
+    DALI_ENFORCE(ws.NumInput() == 3,
+      "Expected 3 inputs. Received: " + std::to_string(ws.NumInput()));
+    const auto &crop_anchor = ws.Input<CPUBackend>(1);
+    const auto &crop_shape = ws.Input<CPUBackend>(2);
+    VerifyArgsShape(crop_anchor.shape(), crop_shape.shape());
+    ProcessArgumentsHelper(ws.data_idx(), crop_anchor.data<float>(), crop_shape.data<float>());
+  }
 
   const CropWindowGenerator& GetCropWindowGenerator(std::size_t data_idx) const {
     DALI_ENFORCE(data_idx < crop_window_generators_.size());
     return crop_window_generators_[data_idx];
   }
 
-  std::vector<float> crop_height_;
-  std::vector<float> crop_width_;
-  std::vector<float> crop_x_norm_;
-  std::vector<float> crop_y_norm_;
-  std::vector<CropWindowGenerator> crop_window_generators_;
-
  private:
-  void ProcessArgumentsHelper(int data_idx, float crop_w, float crop_h,
-                              float crop_x_norm, float crop_y_norm) {
-    crop_width_[data_idx] = crop_w;
-    crop_height_[data_idx] = crop_h;
-    crop_x_norm_[data_idx] = crop_x_norm;
-    crop_y_norm_[data_idx] = crop_y_norm;
+  void ProcessArgumentsHelper(int data_idx,
+                              const float *slice_anchor_data,
+                              const float *slice_shape_data) {
+    crop_window_generators_[data_idx] =
+      [this, slice_anchor_data, slice_shape_data](const TensorShape<> &shape,
+                                                  const TensorLayout& shape_layout) {
+        CropWindow slice;
+        slice.anchor = std::vector<int64_t>(shape.size(), 0);
+        slice.shape = shape;
 
-    DALI_ENFORCE(crop_x_norm + crop_w <= 1.0f,
-      "crop_x[" + std::to_string(crop_x_norm) + "] + crop_width["
-      + std::to_string(crop_w) + "] must be <= 1.0f");
-    DALI_ENFORCE(crop_y_norm + crop_h <= 1.0f,
-      "crop_y[" + std::to_string(crop_y_norm) + "] + crop_height["
-      + std::to_string(crop_h) + "] must be <= 1.0f");
+        if (!dim_names_.empty()) {
+          dims_ = {};
```

Comment:
Isn't that a race? On CPU all the threads in thread pool can overwrite `dims_` on their own, clearing it and pushing back to it in a mixed order.

Reply:
You are right. Will fix

Reply:
Done